example/auditor.py

```diff
@@ -0,0 +1,74 @@
+"""
+In this example we setup a simple auditing mechanism that reports information
+about the last channel changed (which channel, when, and who by). Since we
+might need to know this information even when the program is not running we
+persist this data to file, including information about when changes could
+have been made.
+"""
+
+import time
+
+from p4p.nt.scalar import NTScalar
+from p4p.server import Server
+from p4p.server.raw import Handler
+from p4p.server.thread import SharedPV
+
+
+class Auditor(Handler):
+    """Persist information to file so we can audit when the program is closed"""
+
+    def open(self, value):
+        with open("audit.log", mode="a+") as f:
+            f.write(f"Auditing opened at {time.ctime()}\n")
+
+    def close(self, pv):
+        with open("audit.log", mode="a+") as f:
+            value = pv.current().raw["value"]
+            if value:
+                f.write(f"Auditing closed at {time.ctime()}; {value}\n")
+            else:
+                f.write(f"Auditing closed at {time.ctime()}; no changes made\n")
+
+
+class Audited(Handler):
+    """Forward information about Put operations to the auditing PV"""
+
+    def __init__(self, pv: SharedPV):
```

Comment:
I note that this type hint doesn't work in Python2; I don't know how important it is for all the examples to still support Python2, but the overall library still does.